jetty-ee10/jetty-ee10-proxy/src/main/java/org/eclipse/jetty/ee10/proxy/AsyncMiddleManServlet.java

```diff
@@ -610,75 +611,88 @@ public boolean offer(ByteBuffer content, Callback callback)
         {
             if (_log.isDebugEnabled())
                 _log.debug("{} proxying content to downstream: {} bytes {}", getRequestId(clientRequest), content.remaining(), callback);
-            return chunks.offer(new Chunk(content, callback));
+            try (AutoLock ignored = lock.lock())
+            {
+                return chunks.offer(new Chunk(content, callback));
+            }
         }
 
         @Override
-        public void onWritePossible() throws IOException
+        protected Action process() throws Throwable
         {
             ServletOutputStream output = clientRequest.getAsyncContext().getResponse().getOutputStream();
 
-            // If we had a pending write, let's succeed it.
-            if (writePending)
+            Chunk chunk;
+            try (AutoLock ignored = lock.lock())
             {
-                if (_log.isDebugEnabled())
-                    _log.debug("{} pending async write complete of {} on {}", getRequestId(clientRequest), chunk, output);
-                writePending = false;
-                if (succeed(chunk.callback))
-                    return;
+                chunk = this.chunk = chunks.poll();
```

Comment:
You can assert that `this.chunk` is null here.

Reply:
This isn't necessary, but it helps readability. I personally like nulling-out ASAP and asserting when null is expected just to help future maintenance, akin to the methods expecting a lock to be held that have a `isHeldByCurrentThread` assertion.